src/main/java/ui/TextLineNumber.java

```diff
@@ -0,0 +1,292 @@
+package ui;
+import java.awt.*;
+import java.beans.*;
+import java.util.HashMap;
+import javax.swing.*;
+import javax.swing.border.*;
+import javax.swing.event.*;
+import javax.swing.text.*;
+
+
+public class TextLineNumber extends JPanel
+        implements CaretListener, DocumentListener, PropertyChangeListener
+{
+    public final static float LEFT = 0.0f;
+    public final static float CENTER = 0.5f;
+    public final static float RIGHT = 1.0f;
+
+    private final static Border OUTER = new MatteBorder(0, 0, 0, 2, Color.GRAY);
+
+    private final static int HEIGHT = Integer.MAX_VALUE - 1000000;
+
+    private JTextComponent component;
+
+    private boolean updateFont;
+    private int borderGap;
+    private Color currentLineForeground;
+    private float digitAlignment;
+    private int minimumDisplayDigits;
+
+    private int lastDigits;
+    private int lastHeight;
+    private int lastLine;
+
+    private HashMap<String, FontMetrics> fonts;
+
+    public TextLineNumber(JTextComponent component)
+    {
```

Comment:
Optional: In order to use the same coding style. I recommend you put the let curly bracket at the end of the line: That is: 
public TextLineNumber() {

But not: 
public TextLineNumber()
{